### PR TITLE
Include release builds in GitHub Actions

### PR DESCRIPTION
**Pull request by StuartDBrady**
The "Ninja" CMake generator does not support multiple configurations, i.e. it does not support use of the '--config' option when running 'cmake --build'.  As such, the default configuration (i.e. Debug) was getting used for all builds.
    
Use the CMAKE_BUILD_TYPE variable instead, so that we do release builds, but change one build (ubuntu-20.04 aarch64) to use Debug as its build type, to keep some build coverage for asserts, etc.
    
For Vulkan-Loader and OpenCL-ICD-Loader, we do release builds unconditionally, as we assume there is no need in the CI workflow to actually run the binaries that are built, and therefore no need for any additional debug info.

Signed-off-by: Stuart Brady <stuart.brady@arm.com>

### CODE REVIEW

File: presubmit.sh
Comment (kpet) [inline presubmit.sh:43]: Missing $?

Comment (StuartDBrady) [reply presubmit.sh]: Thanks, fixed.

File: .github/workflows/presubmit.yml
Comment (MeyeChris) [inline .github/workflows/presubmit.yml:27]: What does this do?

I'm not conversant with the CI but how does this change stay scoped to just the Debug build?

Comment (StuartDBrady) [reply .github/workflows/presubmit.yml:27]: I'm not sure I really understand the question, but I will try my best to explain.

There are five builds specified here:
* ubuntu-20.04 (x86-64)
* ubuntu-20.04 arm
* ubuntu-20.04 aarch64
* macos-latest
* windows-latest

The "include" section is what causes the CTS to be built with additional architectures for ubuntu-20.04.

"debug: 1" only applies to ubuntu-20.04 with the arch being "aarch64".

The "ubuntu-20.04 arm" and "ubuntu-20.04" (x86-64) builds do not contain "debug: 1" in their configuration.  In these cases, JOB_ENABLE_DEBUG will not be "1".  (We could perhaps add a default of 0, but there is no strict need, and the same would apply for gl - so it seems best to keep them consistent for now.)

See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-expanding-configurations for documentation regarding workflows in GitHub Actions.

Comment (MeyeChris) [reply .github/workflows/presubmit.yml:27]: Thanks you, completely clear now - I should have looked more closely and rtfm.

So in summary we only do a debug build on ubuntu-20.04-aarch64.

This does mean - if I understand correctly - that we have not just added a debug build option, but changed the ubuntu-20.04-aarch64 from release to debug?

Sorry to be a pedant, and I can be :-), but should the commit message not say something along Add debug build option and change ubuntu-20.04-aarch64 from Release to Debug?

File: presubmit.sh
Comment (svenvh) [inline presubmit.sh:93]: I'm not sure if this does what's intended.  Before this change, we'd expect all builds to be release builds, but [this comment](https://github.com/KhronosGroup/OpenCL-CTS/pull/1483#issuecomment-1241886877) makes me think they're actually all debug builds?

As far as I know, the build type is typically set in the configuration step using `-D CMAKE_BUILD_TYPE=Release`.

Comment (StuartDBrady) [reply presubmit.sh:93]: I think you might be right.  It turns out that the "Ninja" generator _does not support_ multiple configurations using `--config` (as listed in CMAKE_CONFIGURATION_TYPES, and defaulting to CMAKE_DEFAULT_BUILD_TYPE if set, or otherwise the first configuration type in the list).  There is a special "Ninja Multi-Config" generator that supports multiple configurations, but it was added in CMake 3.17.  The version of CMake included in Ubuntu 20.04 is 3.16, so it is too old to use "Ninja Multi-Config".

It's probably best if we use `CMAKE_BUILD_TYPE=Release`.

File: .github/workflows/presubmit.yml
Comment (StuartDBrady) [reply .github/workflows/presubmit.yml:27]: That would have been the case, except we've discovered that we have _only_ doing debug builds so far, and that we should include release builds.

"Add debug build option and change ubuntu-20.04-aarch64 from Release to Debug?" might have been a little on the verbose side.  I will go with "Include release builds in GitHub Actions" as the commit title and elaborate below, unless there's a more accurate title I could use that is also concise.

**Comment by svenvh** [comment]
The list of checks for this PR contains the following entry (which looks different from the other Presubmit entries):
```
Build ubuntu-20.04 Expected — Waiting for status to be reported 
```
Not sure if there is something in this change that's causing this stale(?) entry?

**Comment by StuartDBrady** [comment]
> The list of checks for this PR contains the following entry (which looks different from the other Presubmit entries):
> 
> ```
> Build ubuntu-20.04 Expected — Waiting for status to be reported 
> ```
> 
> Not sure if there is something in this change that's causing this stale(?) entry?

Yes, the GitHub config for the KhronosGroup/OpenCL-CTS repo waits for a number of builds to finish, one of which being "ubuntu-20.04".  However, I changed this to "ubuntu-20.04 gl" at the same time as using the name "ubuntu-20.04 aarch64 debug".  Once we merge this change, I can update the GitHub config after this change has been merged.  Alternatively, I could remove the part of the change that sets " gl" as the extra info for the default-arch "ubuntu-20.04" build.

File: .github/workflows/presubmit.yml
Comment (svenvh) [inline .github/workflows/presubmit.yml:6]: What's the reason for not doing
```suggestion
    name: Build ${{ matrix.os }} ${{ matrix.arch }} ${{ matrix.extra }}
```
and avoiding the leading space in the `extra:` specifications below?  I fear the leading space might be easily overlooked/forgotten.

**Comment by svenvh** [comment]
> Yes, the GitHub config for the KhronosGroup/OpenCL-CTS repo waits for a number of builds to finish, one of which being "ubuntu-20.04". However, I changed this to "ubuntu-20.04 gl" at the same time as using the name "ubuntu-20.04 aarch64 debug". Once we merge this change, I can update the GitHub config after this change has been merged.

Okay, if this is expected then that sounds okay to me.

File: .github/workflows/presubmit.yml
Comment (StuartDBrady) [reply .github/workflows/presubmit.yml:6]: To avoid including a trailing space in the name where there matrix.extra is unset.

**Comment by svenvh** [comment]
LGTM, but would be good to have approval from someone else.

**Review by kpet — APPROVED** [review]
LGTM!

@StuartDBrady If you want to keep working on the CI, you might want to consider continuing the work I started to get rid of `presubmit.sh`.

**Pull request merged**
